client/src/entrypoints/admin/draftail.js

```diff
@@ -11,52 +11,59 @@ import draftail, {
-window.chooserUrls = global.chooserUrls || {};
-
-// Plugins for the built-in entities.
-const entityTypes = [
-  {
-    type: 'DOCUMENT',
-    source: draftail.DocumentModalWorkflowSource,
-    decorator: Document,
-  },
-  {
-    type: 'LINK',
-    source: draftail.LinkModalWorkflowSource,
-    decorator: Link,
-    onPaste: onPasteLink,
-  },
-  {
-    type: 'IMAGE',
-    source: draftail.ImageModalWorkflowSource,
-    block: ImageBlock,
-  },
-  {
-    type: 'EMBED',
-    source: draftail.EmbedModalWorkflowSource,
-    block: EmbedBlock,
-  },
-];
-
-entityTypes.forEach((type) => draftail.registerPlugin(type, 'entityTypes'));
+// This file is included and run when there's a DraftailRichTextArea widget in the response.
+// Normally this is only included once in the initial page load, but it may be included
+// more than once when there's an AJAX response that includes the widget, e.g. in choosers.
+// Ensure we only run the initialization code once.
+// https://github.com/wagtail/wagtail/issues/12002
+if (!window.Draftail || !window.draftail) {
+  // Expose Draftail package as a global.
+  window.Draftail = Draftail;
+  // Expose module as a global.
+  window.draftail = draftail;
 
-/**
- * Initialize a Draftail editor on a given element when the w-draftail:init event is fired.
- */
-document.addEventListener('w-draftail:init', ({ detail = {}, target }) => {
-  const id = target.id;
+  /** @deprecated RemovedInWagtail70 - Ensure that any third party packages that use global.chooserUrls can still append to this global object until support is removed. */
+  window.chooserUrls = global.chooserUrls || {};
+
+  // Plugins for the built-in entities.
+  const entityTypes = [
+    {
+      type: 'DOCUMENT',
+      source: draftail.DocumentModalWorkflowSource,
+      decorator: Document,
+    },
+    {
+      type: 'LINK',
+      source: draftail.LinkModalWorkflowSource,
+      decorator: Link,
+      onPaste: onPasteLink,
+    },
+    {
+      type: 'IMAGE',
+      source: draftail.ImageModalWorkflowSource,
+      block: ImageBlock,
+    },
+    {
+      type: 'EMBED',
+      source: draftail.EmbedModalWorkflowSource,
+      block: EmbedBlock,
+    },
+  ];
+
+  entityTypes.forEach((type) => draftail.registerPlugin(type, 'entityTypes'));
+
+  /**
+   * Initialize a Draftail editor on a given element when the w-draftail:init event is fired.
+   */
+  document.addEventListener('w-draftail:init', ({ detail = {}, target }) => {
+    const id = target.id;
 
-  if (!id) {
-    // eslint-disable-next-line no-console
-    console.error('`w-draftail:init` event must have a target with an id.');
-    return;
-  }
+    if (!id) {
+      // eslint-disable-next-line no-console
+      console.error('`w-draftail:init` event must have a target with an id.');
+      return;
+    }
 
-  window.draftail.initEditor(`#${id}`, detail, document.currentScript);
-});
+    window.draftail.initEditor(`#${id}`, detail, document.currentScript);
```

Comment:
Only the `addEventListener` needs to be run once, but I think we can save some unnecessary computation by wrapping the entire script in the `if` check.